File: backend/services/blob_store.py
"""Blob 存储服务 - 集中管理内存中的二进制资源"""
from __future__ import annotations

import hashlib
import mimetypes
from typing import Dict, Tuple, Callable

# 键: SHA256 哈希, 值: (字节数据, MIME 类型, 扩展名)
blob_store: Dict[str, Tuple[bytes, str, str]] = {}


def compute_sha256(data: bytes) -> str:
    """计算字节数据的 SHA-256 哈希"""
    return hashlib.sha256(data).hexdigest()


def get_mime_type(ext: str) -> str:
    """根据扩展名返回 MIME 类型"""
    mime_map = {
        '.png': 'image/png',
        '.jpg': 'image/jpeg',
        '.jpeg': 'image/jpeg',
        '.gif': 'image/gif',
        '.webp': 'image/webp',
        '.bmp': 'image/bmp',
        '.tif': 'image/tiff',
        '.tiff': 'image/tiff',
        '.svg': 'image/svg+xml',
    }
    return mime_map.get(ext.lower(), 'application/octet-stream')


def store_blob(data: bytes, ext: str = '.png') -> str:
    """
    存储字节数据到内存 blob 存储，返回其 SHA256 哈希值

    重复的内容会自动去重（相同数据返回相同哈希）
    """
    content_hash = compute_sha256(data)
    if content_hash not in blob_store:
        mime = get_mime_type(ext)
        blob_store[content_hash] = (data, mime, ext)
        print(f"[blob_store] 存储 blob: {content_hash[:12]}... (大小: {len(data)} 字节, MIME: {mime})")
    else:
        print(f"[blob_store] blob 已存在（去重）: {content_hash[:12]}...")
    return content_hash


def get_blob(blob_hash: str) -> Tuple[bytes, str, str] | None:
    """从 blob 存储中获取数据"""
    return blob_store.get(blob_hash)


def clear_blob_store() -> None:
    """清空 blob 存储（仅用于测试）"""
    blob_store.clear()


def get_store_size() -> int:
    """返回存储中的 blob 数量"""
    return len(blob_store)

File: backend-node/src/services/blobStore.js
/**
 * Blob 存储服务 - 内存存储图片数据
 */
import crypto from 'crypto';
import mime from 'mime-types';

// 内存存储：Map<hash, { data: Buffer, mime: string, ext: string }>
const blobStore = new Map();

/**
 * 计算数据的 SHA256 哈希
 */
export function computeSHA256(data) {
  return crypto.createHash('sha256').update(data).digest('hex');
}

/**
 * 根据扩展名获取 MIME 类型
 */
export function getMimeType(ext) {
  const mimeMap = {
    '.png': 'image/png',
    '.jpg': 'image/jpeg',
    '.jpeg': 'image/jpeg',
    '.gif': 'image/gif',
    '.webp': 'image/webp',
    '.bmp': 'image/bmp',
    '.tif': 'image/tiff',
    '.tiff': 'image/tiff',
    '.svg': 'image/svg+xml',
  };
  return mimeMap[ext.toLowerCase()] || 'application/octet-stream';
}

/**
 * 存储 blob 数据，返回哈希值
 */
export function storeBlob(data, ext = '.png') {
  const hash = computeSHA256(data);
  
  if (!blobStore.has(hash)) {
    const mimeType = getMimeType(ext);
    blobStore.set(hash, {
      data,
      mime: mimeType,
      ext,
    });
    console.log(`[blob_store] 存储 blob: ${hash.slice(0, 12)}... (大小: ${data.length} 字节, MIME: ${mimeType})`);
  } else {
    console.log(`[blob_store] blob 已存在（去重）: ${hash.slice(0, 12)}...`);
  }
  
  return hash;
}

/**
 * 获取 blob 数据
 */
export function getBlob(hash) {
  return blobStore.get(hash);
}

/**
 * 清空存储
 */
export function clearBlobStore() {
  blobStore.clear();
}

/**
 * 获取存储大小
 */
export function getStoreSize() {
  return blobStore.size;
}
